cmds/wget/wget.go

```diff
@@ -37,6 +43,13 @@ func wget(arg string, w io.Writer) error {
 	if resp.StatusCode != 200 {
 		return fmt.Errorf("non-200 HTTP status: %d", resp.StatusCode)
 	}
+
+	w, err := os.Create(fileName)
+	if err != nil {
+		return err
+	}
+	defer w.Close()
```

Comment:
There are warnings in various places about using defer when you do not need to. There's no guarantee about *when* the close will happen. I've been bitten by this in real life when opening thousands of files too quickly. In this case, since you do a copy and immediately return, you might consider doing the close in line. I realize it does not matter here, for one file, but ...

Reply:
This is almost the canonical use case for defer - no problem with it. There are only two possible downsides to a defer: you don't get the error, and the performance is worse than manually adding the call to each return path. Neither, I think, are particularly important in this case.

Reply:
Sure. but see this: https://joeshaw.org/dont-defer-close-on-writable-files/

and be aware that sometimes defer on close can bite you. It has bitten me among others. If you really care about the file you are writing, deferring a close can be a bad way to go.

It's not important for this instance I suppose, save that we have another case of ignoring an error from close when we don't have to (and, yes, you can get an error on close).

Reply:
You are correct, close can actually return errors (and in this case, maybe important ones, in case writeback to the actual device fails). That's not a problem with defer in general, though.

On the other hand, if Close returns an error, there isn't much the user application can do about it anyway.

Reply:
Actually, you're right, we should probably check Close if io.Copy proceeds. You could do what that guy suggested actually:

```go
defer f.Close()

if _, err := io.Copy(..); err != nil {
  return err
}
return f.Close()
```

since Close is idempotent, you can call it twice, and the deferred close will be called after the `f.Close()` in return is called.

Reply:
So as a simple example, Unix close used to only return EBADFD. 
See http://man.cat-v.org/unix-6th/2/close
In Unix, you could not lose data on a close; all it really did was zero the file table in the proc struct and decrement a reference count. Not much to it.

Once we had buffer caches and deferred writes, it became possible for close to return an error, but many programs had been written to not check since EBADFD was not something easily fixed. So what error might it return?  It's hardly consistent, here's one from freebsd:
[ENOSPC]		The underlying object did not fit, cached data was lost.

Linux has different errors, as does OSX.

People were surprised back in the 80s when this became a thing and they exited an editor to find that, not only had the file not been changed, it was actually zero length, as the first buffer cache component written at zero truncated the file but the write itself failed. Oh boy. 

Is it true the program can do nothing? in the case of editors they can do a lot. They can warn you and refuse to exit. That gives you a chance to save your data some other way, maybe by removing a file so you can write the file you want.

In many cases, defer'ed close of a writeable file is asking for trouble. There are ways to recover. In the case of this program, where we might just do the wget again, the code you show above is quite nice and we should use it. At least we know if failed. We should look at our uses of defer'ed close on writeable files to see if there are ways to use your suggestion.